airflow/utils/pydantic.py

```diff
@@ -0,0 +1,58 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# This is an util module that makes Pydantic use optional. While we are using Pydantic in the airflow core
+# codebase, we don't want to make it a hard dependency for all the users of the core codebase, because
+# it is only used in the serialization and deserialization of the models for Internal API and for nothing
+# else, and since Pydantic is a very popular library, we don't want to force the users of the core codebase
+# to install specific Pydantic version - especially that a lot of libraries out there still depend on
+# Pydantic 1 and our internal API uses Pydantic 2+
+
+from __future__ import annotations
+
+try:
+    import pydantic
+    from packaging.version import Version
+
+    if Version(pydantic.__version__) < Version("2.0.0"):
+        raise ImportError("Pydantic version is too old. Switch to Pydantic 2+")
```

Comment:
I guess you would hate me if it would start to talk about `__version__` again 🤣 

Maybe we should consider to switch to importlib? I've compare locally and with `pydantic` retrieve the version from the importlib suddenly take less time if compare to `pydantic.__version__`. (Python 3.9, macOS, pydantic 2.4.2)

```python
import sys
if sys.version_info >= (3, 9):
    from importlib.metadata import distribution
else:
    from importlib_metadata import distribution

try:
    pydantic_distribution = distribution("pydantic")
    pydantic_distribution.version
except ImportError:
    pass
```

It took from 9 to 15 ms, during the run in PyCharm Profiler. Regardless is it installed or not
![image](https://github.com/apache/airflow/assets/3998685/5764d241-8da4-4d4c-8a67-9fb418dcf370)

If compare to 

try:
    import pydantic
    pydantic.__version__
except ImportError:
    pass

It took from 43 to 60 ms if `pydantic` installed, and 0-1 ms if not.
It is also depend on the version, 1.x took longer than 2.4.x, and 2.6.1 (latest) import fast due to lazy loading into the `__init__.py`
![image](https://github.com/apache/airflow/assets/3998685/a326dd9c-d826-4db2-a7f5-78e377736441)

I know this is not such a big different, if compare to Airflow loading time

Reply:
Fixed.

Reply:
Depends on. New versions might be faster rather than `importlib`, but old one slower.

Reply:
And also most of the time spend to import `importlib`, if we already imported it should spend less time to the run `importlib.metadata.distribution`

Reply:
Well. in this case it will be simply faster because we won't really even need to import pydantic in this case - and that  will be quite a saviing - both time and memory if Pydantic 1 is installed (because we will not need it). Also it avoids potential side-effects of importing Pydantic in general, so in this case (unlike reading`airflow.__version__`  from providers) it is way better.